docs/madr/decisions/042-tcp-grpc-probe.md

```diff
@@ -0,0 +1,246 @@
+      tcpSocket:
+        host: some-other-host
+        port: 1433
+```
+
+The `host` field in the `tcpSocket` probe is only used to resolve the network address when one wants to probe a different host than the pod's IP address and it's optional. The `service` field on gRPC probes is also optional. There is little space for us to carry contextual information on them.
+
+A pod can have multiple containers with each defines multiple and different probes, the worst case is a pod has all 3 different types of probes with many probes for each type.
+
+## Existing design: HTTP Virtual Probes
+
+As per our current design, a pod with multiple probes will be merged and will be handled by a single virtual listener. This listener will forward/redirect probe requests to the corresponding application port without any intermediate transforming except path rewriting, etc.
+
+While we provided annotations for users to customize whether they want to enable the Virtual Probe feature and on which port do they want to expose the Virtual Probe listener, the current design actually does not require the user to customize these items and the probes can be translated automatically and transparently.
+
+The following example demonstrates how HTTP probes are handled in current implementation. Two probes are both handled by the Virtual Probe listener on port 19000:
+
+[inject.14.input.yaml](https://github.com/kumahq/kuma/blob/master/pkg/plugins/runtime/k8s/webhooks/injector/testdata/inject.14.input.yaml)
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: busybox
+  labels:
+    run: busybox
+  annotations:
+    kuma.io/virtual-probes-port: "19000"
+spec:
+  containers:
+    - name: busybox
+      image: busybox
+      resources: {}
+      readinessProbe:
+        httpGet:
+          path: /metrics
+          port: 3001
+        initialDelaySeconds: 3
+        periodSeconds: 3
+      livenessProbe:
+        httpGet:
+          path: /metrics
+          port: 8080
+        initialDelaySeconds: 3
+        periodSeconds: 3
+```
+
+[inject.14.golden.yaml](https://github.com/kumahq/kuma/blob/master/pkg/plugins/runtime/k8s/webhooks/injector/testdata/inject.14.golden.yaml)
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    kuma.io/virtual-probes: enabled
+    kuma.io/virtual-probes-port: "19000"
+    ...
+  labels:
+    run: busybox
+  name: busybox
+spec:
+  containers:
+  - name: kuma-sidecar
+    image: kuma/kuma-sidecar:latest
+    ...
+  - image: busybox
+    livenessProbe:
+      httpGet:
+        path: /8080/metrics
+        port: 19000
+      initialDelaySeconds: 3
+      periodSeconds: 3
+    name: busybox
+    readinessProbe:
+      httpGet:
+        path: /3001/metrics
+        port: 19000
+      initialDelaySeconds: 3
+      periodSeconds: 3
+  initContainers:
+  - name: kuma-init
+    image: kuma/kuma-init:latest
+  ...
+```
+
+## Considered Options
+
+- Allocate a new and separated port for each of the TCP probes and gRPC probes
```

Comment:
Why do gRPC probes need separate ports? AFAIU gRPC is just HTTP so it could just be a different route on the http probe no?

Reply:
Multiple gRPC probes to different ports will use the same path and method, so we can't differentiate them within one single listener.

Reply:
Is this a realistic use case? Can we find an example where we'd end up in this case?

Reply:
We can also make it support only one gRPC probe per pod, if that makes sense

Reply:
It turns out that we can support mulple gRPC probes on a single port.